File: web/status-history/src/test/java/net/twisterrob/blt/gapp/DatastoreStatusHistoryRepositoryUnitTest.java
package net.twisterrob.blt.gapp;

import java.util.Collections;
import java.util.Iterator;
import java.util.List;
import java.util.Map;
import java.util.UUID;

import org.junit.Test;
import org.mockito.ArgumentCaptor;
import org.mockito.ArgumentMatchers;

import static org.hamcrest.MatcherAssert.assertThat;
import static org.hamcrest.Matchers.hasEntry;
import static org.hamcrest.Matchers.hasSize;
import static org.hamcrest.Matchers.instanceOf;
import static org.junit.Assert.assertEquals;
import static org.mockito.Mockito.doAnswer;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

import com.google.cloud.Timestamp;
import com.google.cloud.datastore.BaseKey;
import com.google.cloud.datastore.Datastore;
import com.google.cloud.datastore.Entity;
import com.google.cloud.datastore.EntityQuery;
import com.google.cloud.datastore.FullEntity;
import com.google.cloud.datastore.Key;
import com.google.cloud.datastore.KeyFactory;
import com.google.cloud.datastore.QueryResults;
import com.google.cloud.datastore.StringValue;
import com.google.cloud.datastore.StructuredQuery.OrderBy;
import com.google.cloud.datastore.TimestampValue;
import com.google.cloud.datastore.Value;

import kotlin.random.Random;
import kotlinx.datetime.Instant;

import net.twisterrob.blt.gapp.infrastructure.DatastoreStatusHistoryRepository;
import net.twisterrob.travel.domain.london.status.Feed;
import net.twisterrob.travel.domain.london.status.Stacktrace;
import net.twisterrob.travel.domain.london.status.StatusContent;
import net.twisterrob.travel.domain.london.status.StatusItem;
import net.twisterrob.travel.domain.london.status.StatusItem.FailedStatusItem;
import net.twisterrob.travel.domain.london.status.StatusItem.SuccessfulStatusItem;
import net.twisterrob.travel.domain.london.status.api.StatusHistoryRepository;

public class DatastoreStatusHistoryRepositoryUnitTest {

	private final Datastore mockDatastore = mock(Datastore.class);
	private final StatusHistoryRepository subject = new DatastoreStatusHistoryRepository(mockDatastore);

	@Test public void testSavesSuccessfulItem() {
		when(mockDatastore.newKeyFactory()).thenReturn(new KeyFactory("test-project-id"));

		SuccessfulStatusItem item = SuccessfulStatusItem();

		subject.add(item);

		FullEntity<? extends BaseKey> entity = captureAddedEntity(mockDatastore);
		assertEquals(item.getFeed().name(), entity.getKey().getKind());
		Map<String, Value<?>> props = entity.getProperties();
		assertThat(props.entrySet(), hasSize(2));
		Instant now = item.getRetrievedDate();
		Timestamp ts = Timestamp.ofTimeSecondsAndNanos(now.getEpochSeconds(), now.getNanosecondsOfSecond());
		assertThat(props, hasEntry("retrievedDate", TimestampValue.of(ts)));
		String content = item.getContent().getContent();
		assertThat(props, hasEntry("content", StringValue.newBuilder(content).setExcludeFromIndexes(true).build()));
	}

	@Test public void testSavesFailedItem() {
		when(mockDatastore.newKeyFactory()).thenReturn(new KeyFactory("test-project-id"));
		FailedStatusItem item = FailedStatusItem();

		subject.add(item);

		FullEntity<? extends BaseKey> entity = captureAddedEntity(mockDatastore);
		assertEquals(item.getFeed().name(), entity.getKey().getKind());
		Map<String, Value<?>> props = entity.getProperties();
		assertThat(props.entrySet(), hasSize(2));
		Instant now = item.getRetrievedDate();
		Timestamp ts = Timestamp.ofTimeSecondsAndNanos(now.getEpochSeconds(), now.getNanosecondsOfSecond());
		assertThat(props, hasEntry("retrievedDate", TimestampValue.of(ts)));
		String error = item.getError().getStacktrace();
		assertThat(props, hasEntry("error", StringValue.newBuilder(error).setExcludeFromIndexes(true).build()));
	}

	@Test public void testQuery() {
		@SuppressWarnings("unchecked")
		QueryResults<Entity> mockResults = mock(QueryResults.class);
		doAnswerIterator(mockResults, Collections.emptyIterator());
		when(mockDatastore.run(ArgumentMatchers.<EntityQuery>any())).thenReturn(mockResults);

		Feed feed = randomFeed();
		subject.getAll(feed, 2);

		EntityQuery query = captureRanQuery(mockDatastore);
		assertEquals(feed.name(), query.getKind());
		assertEquals(List.of(OrderBy.desc("retrievedDate")), query.getOrderBy());
		assertEquals(Integer.valueOf(2), query.getLimit());
	}

	@Test public void testLimitShortensLongerList() {
		List<Entity> list = List.of(SuccessfulEntity(), SuccessfulEntity(), SuccessfulEntity());
		@SuppressWarnings("unchecked")
		QueryResults<Entity> mockResults = mock(QueryResults.class);
		doAnswerIterator(mockResults, list.iterator());
		when(mockDatastore.run(ArgumentMatchers.<EntityQuery>any())).thenReturn(mockResults);

		List<StatusItem> result = subject.getAll(randomFeed(), 2);

		assertThat(result, hasSize(2));
	}

	@Test public void testReturnsSuccessfulEntityData() {
		Feed feed = randomFeed();
		Entity entity = SuccessfulEntity(feed);
		@SuppressWarnings("unchecked")
		QueryResults<Entity> mockResults = mock(QueryResults.class);
		List<Entity> list = List.of(entity);
		doAnswerIterator(mockResults, list.iterator());
		when(mockDatastore.run(ArgumentMatchers.<EntityQuery>any())).thenReturn(mockResults);

		List<StatusItem> result = subject.getAll(randomFeed(), 2);

		StatusItem item = result.get(0);
		assertEquals(feed, item.getFeed());
		Timestamp ts = entity.getTimestamp("retrievedDate");
		Instant instant = Instant.Companion.fromEpochSeconds(ts.getSeconds(), ts.getNanos());
		assertEquals(instant, item.getRetrievedDate());
		assertThat(item, instanceOf(SuccessfulStatusItem.class));
		assertEquals(entity.getString("content"), ((SuccessfulStatusItem)item).getContent().getContent());
	}

	@Test public void testReturnsFailedEntityData() {
		Feed feed = randomFeed();
		Entity entity = FailedEntity(feed);
		List<Entity> list = List.of(entity);
		@SuppressWarnings("unchecked")
		QueryResults<Entity> mockResults = mock(QueryResults.class);
		doAnswerIterator(mockResults, list.iterator());
		when(mockDatastore.run(ArgumentMatchers.<EntityQuery>any())).thenReturn(mockResults);

		List<StatusItem> result = subject.getAll(randomFeed(), 2);

		StatusItem item = result.get(0);
		assertEquals(feed, item.getFeed());
		Timestamp ts = entity.getTimestamp("retrievedDate");
		Instant instant = Instant.Companion.fromEpochSeconds(ts.getSeconds(), ts.getNanos());
		assertEquals(instant, item.getRetrievedDate());
		assertThat(item, instanceOf(FailedStatusItem.class));
		assertEquals(entity.getString("error"), ((FailedStatusItem)item).getError().getStacktrace());
	}

	@Test public void testSymmetricSaveLoadSuccessfulItem() {
		SuccessfulStatusItem item = SuccessfulStatusItem();
		when(mockDatastore.newKeyFactory()).thenReturn(new KeyFactory("test-project-id"));
		subject.add(item);
		FullEntity<? extends BaseKey> savedEntity = captureAddedEntity(mockDatastore);
		Entity entity = Entity.newBuilder(randomKey(item.getFeed().name()), savedEntity).build();

		@SuppressWarnings("unchecked")
		QueryResults<Entity> mockResults = mock(QueryResults.class);
		doAnswerIterator(mockResults, List.of(entity).iterator());
		when(mockDatastore.run(ArgumentMatchers.<EntityQuery>any())).thenReturn(mockResults);

		List<StatusItem> result = subject.getAll(randomFeed(), 1);
		assertEquals(item, result.get(0));
	}

	private static <T> void doAnswerIterator(Iterator<? super T> mock, Iterator<T> value) {
		doAnswer(invocation -> value.next()).when(mock).next();
		doAnswer(invocation -> value.hasNext()).when(mock).hasNext();
	}

	private static EntityQuery captureRanQuery(Datastore mockDatastore) {
		ArgumentCaptor<EntityQuery> captor = ArgumentCaptor.forClass(EntityQuery.class);
		verify(mockDatastore).run(captor.capture());
		assertThat(captor.getAllValues(), hasSize(1));
		return captor.getAllValues().get(0);
	}

	private static FullEntity<? extends BaseKey> captureAddedEntity(Datastore mockDatastore) {
		@SuppressWarnings("unchecked")
		ArgumentCaptor<FullEntity<? extends BaseKey>> captor = ArgumentCaptor.forClass(FullEntity.class);
		verify(mockDatastore).add(captor.capture());
		assertThat(captor.getAllValues(), hasSize(1));
		return captor.getAllValues().get(0);
	}

	private static Entity SuccessfulEntity() {
		return SuccessfulEntity(randomFeed());
	}

	private static Entity SuccessfulEntity(Feed feed) {
		return Entity
				.newBuilder(randomKey(feed.name()))
				.set("retrievedDate", randomTimestamp())
				.set("content", UUID.randomUUID().toString())
				.build();
	}

	private static Entity FailedEntity(Feed feed) {
		return Entity
				.newBuilder(randomKey(feed.name()))
				.set("retrievedDate", randomTimestamp())
				.set("error", UUID.randomUUID().toString())
				.build();
	}

	private static StatusItem.SuccessfulStatusItem SuccessfulStatusItem() {
		return new StatusItem.SuccessfulStatusItem(
				randomFeed(),
				new StatusContent(UUID.randomUUID().toString()),
				randomInstant()
		);
	}

	private static StatusItem.FailedStatusItem FailedStatusItem() {
		return new StatusItem.FailedStatusItem(
				randomFeed(),
				new Stacktrace(UUID.randomUUID().toString()),
				randomInstant()
		);
	}

	private static Feed randomFeed() {
		return Feed.getEntries().get(Random.Default.nextInt(Feed.getEntries().size()));
	}

	private static Key randomKey(String kind) {
		KeyFactory keyFactory = new KeyFactory("test-project-id").setKind(kind);
		return keyFactory.newKey(UUID.randomUUID().toString());
	}

	private static Timestamp randomTimestamp() {
		return Timestamp.ofTimeSecondsAndNanos(
				Random.Default.nextLong(
						Timestamp.MIN_VALUE.getSeconds(),
						Timestamp.MAX_VALUE.getSeconds()
				),
				Random.Default.nextInt(
						Timestamp.MIN_VALUE.getNanos(),
						Timestamp.MAX_VALUE.getNanos()
				)
		);
	}

	private static Instant randomInstant() {
		System.out.println(Instant.Companion.getDISTANT_PAST());
		System.out.println(Instant.Companion.getDISTANT_FUTURE());
		System.out.println(Timestamp.MIN_VALUE);
		System.out.println(Timestamp.MAX_VALUE);
		// Note: Have to use the range from Google's Timestamp, because Instant.DISTANT_PAST/DISTANT_FUTURE
		// are out of range (-100001-12-31T23:59:59.999999999Z - +100000-01-01T00:00:00Z),
		// compared to Timestamp's 0001-01-01T00:00:00Z - 9999-12-31T23:59:59.999999999Z range.
		return Instant.Companion.fromEpochSeconds(
				Random.Default.nextLong(
						Timestamp.MIN_VALUE.getSeconds(),
						Timestamp.MAX_VALUE.getSeconds()
				),
				Random.Default.nextInt(
						Timestamp.MIN_VALUE.getNanos(),
						Timestamp.MAX_VALUE.getNanos()
				)
		);
	}
}

File: web/status-history/src/test/java/net/twisterrob/travel/statushistory/infrastructure/DatastoreStatusHistoryRepositoryUnitTest.kt
package net.twisterrob.travel.statushistory.infrastructure

import com.google.cloud.Timestamp
import com.google.cloud.datastore.Datastore
import com.google.cloud.datastore.Entity
import com.google.cloud.datastore.EntityQuery
import com.google.cloud.datastore.FullEntity
import com.google.cloud.datastore.KeyFactory
import com.google.cloud.datastore.QueryResults
import com.google.cloud.datastore.StringValue
import com.google.cloud.datastore.StructuredQuery
import com.google.cloud.datastore.TimestampValue
import net.twisterrob.travel.domain.london.status.StatusItem.FailedStatusItem
import net.twisterrob.travel.domain.london.status.StatusItem.SuccessfulStatusItem
import net.twisterrob.travel.domain.london.status.api.StatusHistoryDataSource
import net.twisterrob.travel.statushistory.test.FailedEntity
import net.twisterrob.travel.statushistory.test.FailedStatusItem
import net.twisterrob.travel.statushistory.test.SuccessfulEntity
import net.twisterrob.travel.statushistory.test.SuccessfulStatusItem
import net.twisterrob.travel.statushistory.test.randomFeed
import net.twisterrob.travel.statushistory.test.randomKey
import org.hamcrest.MatcherAssert.assertThat
import org.hamcrest.Matchers.hasEntry
import org.hamcrest.Matchers.hasSize
import org.hamcrest.Matchers.instanceOf
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Test
import org.mockito.ArgumentCaptor
import org.mockito.Mockito.any
import org.mockito.Mockito.doAnswer
import org.mockito.Mockito.mock
import org.mockito.Mockito.spy
import org.mockito.Mockito.times
import org.mockito.Mockito.verify
import org.mockito.Mockito.`when`
import java.util.Collections.emptyIterator
import kotlin.time.Instant

class DatastoreStatusHistoryRepositoryUnitTest {

	private val mockDatastore: Datastore = mock()
	private val subject: StatusHistoryDataSource = DatastoreStatusHistoryDataSource(mockDatastore)

	@Test fun testSavesSuccessfulItem() {
		`when`(mockDatastore.newKeyFactory()).thenReturn(KeyFactory("test-project-id"))

		val item = SuccessfulStatusItem()

		subject.add(item)

		val entity: FullEntity<*> = captureSingle { verify(mockDatastore).add(capture()) }
		assertEquals(item.feed.name, entity.key!!.kind)
		val props = entity.properties
		assertThat(props.entries, hasSize(2))
		val now = item.retrievedDate
		val ts = Timestamp.ofTimeSecondsAndNanos(now.epochSeconds, now.nanosecondsOfSecond)
		assertThat(props, hasEntry("retrievedDate", TimestampValue.of(ts)))
		val content = item.content.content
		assertThat(props, hasEntry("content", StringValue.newBuilder(content).setExcludeFromIndexes(true).build()))
	}

	@Test fun testSavesFailedItem() {
		`when`(mockDatastore.newKeyFactory()).thenReturn(KeyFactory("test-project-id"))
		val item = FailedStatusItem()

		subject.add(item)

		val entity: FullEntity<*> = captureSingle { verify(mockDatastore).add(capture()) }
		assertEquals(item.feed.name, entity.key!!.kind)
		val props = entity.properties
		assertThat(props.entries, hasSize(2))
		val now = item.retrievedDate
		val ts = Timestamp.ofTimeSecondsAndNanos(now.epochSeconds, now.nanosecondsOfSecond)
		assertThat(props, hasEntry("retrievedDate", TimestampValue.of(ts)))
		val error = item.error.stacktrace
		assertThat(props, hasEntry("error", StringValue.newBuilder(error).setExcludeFromIndexes(true).build()))
	}

	@Test fun testQuery() {
		val mockResults: QueryResults<Entity> = mock()
		doAnswerIterator(mockResults, emptyIterator())
		`when`(mockDatastore.run(any<EntityQuery>())).thenReturn(mockResults)

		val feed = randomFeed()
		subject.getAll(feed, 2)

		val query: EntityQuery = captureSingle { verify(mockDatastore).run(capture()) }
		assertEquals(feed.name, query.kind)
		assertEquals(listOf(StructuredQuery.OrderBy.desc("retrievedDate")), query.orderBy)
		assertEquals(2, query.limit)
	}

	@Test fun testLimitShortensLongerList() {
		val list = listOf(SuccessfulEntity(), SuccessfulEntity(), SuccessfulEntity())
		val mockResults: QueryResults<Entity> = mock()
		val iterator = spy(list.iterator())
		doAnswerIterator(mockResults, iterator)
		`when`(mockDatastore.run(any<EntityQuery>())).thenReturn(mockResults)

		val result = subject.getAll(randomFeed(), 2)

		assertThat(result, hasSize(2))
		verify(iterator, times(2)).next()
	}

	@Test fun testReturnsSuccessfulEntityData() {
		val feed = randomFeed()
		val entity = SuccessfulEntity(feed)
		val mockResults: QueryResults<Entity> = mock()
		val list = listOf(entity)
		doAnswerIterator(mockResults, list.iterator())
		`when`(mockDatastore.run(any<EntityQuery>())).thenReturn(mockResults)

		val result = subject.getAll(randomFeed(), 2)

		val item = result[0]
		assertEquals(feed, item.feed)
		val ts = entity.getTimestamp("retrievedDate")
		val instant = Instant.fromEpochSeconds(ts.seconds, ts.nanos)
		assertEquals(instant, item.retrievedDate)
		assertThat(item, instanceOf(SuccessfulStatusItem::class.java))
		assertEquals(entity.getString("content"), (item as SuccessfulStatusItem).content.content)
	}

	@Test fun testReturnsFailedEntityData() {
		val feed = randomFeed()
		val entity = FailedEntity(feed)
		val list = listOf(entity)
		val mockResults: QueryResults<Entity> = mock()
		doAnswerIterator(mockResults, list.iterator())
		`when`(mockDatastore.run(any<EntityQuery>())).thenReturn(mockResults)

		val result = subject.getAll(randomFeed(), 2)

		val item = result[0]
		assertEquals(feed, item.feed)
		val ts = entity.getTimestamp("retrievedDate")
		val instant = Instant.fromEpochSeconds(ts.seconds, ts.nanos)
		assertEquals(instant, item.retrievedDate)
		assertThat(item, instanceOf(FailedStatusItem::class.java))
		assertEquals(entity.getString("error"), (item as FailedStatusItem).error.stacktrace)
	}

	@Test fun testSymmetricSaveLoadSuccessfulItem() {
		val item = SuccessfulStatusItem()
		`when`(mockDatastore.newKeyFactory()).thenReturn(KeyFactory("test-project-id"))
		subject.add(item)
		val savedEntity: FullEntity<*> = captureSingle { verify(mockDatastore).add(capture()) }
		val entity = Entity.newBuilder(randomKey(item.feed.name), savedEntity).build()

		val mockResults: QueryResults<Entity> = mock()
		doAnswerIterator(mockResults, listOf(entity).iterator())
		`when`(mockDatastore.run(any<EntityQuery>())).thenReturn(mockResults)

		val result = subject.getAll(randomFeed(), 1)
		assertEquals(item, result[0])
	}

	companion object {

		private fun <T : Any> doAnswerIterator(mock: Iterator<T>, value: Iterator<T>) {
			doAnswer { value.next() }.`when`(mock).next()
			doAnswer { value.hasNext() }.`when`(mock).hasNext()
		}

		private fun <T> captureSingle(block: ArgumentCaptor<T>.() -> Unit): T {
			val captor: ArgumentCaptor<T> = ArgumentCaptor.captor()
			block(captor)
			assertThat(captor.allValues, hasSize(1))
			return captor.allValues.single()
		}
	}
}
